pallets/omnipool/src/lib.rs

```diff
@@ -2017,18 +2124,48 @@ impl<T: Config> Pallet<T> {
 	}
 
 	/// Calls `on_trade_fee` hook and ensures that no more than the fee amount is transferred.
-	fn process_trade_fee(trader: &T::AccountId, asset: T::AssetId, amount: Balance) -> DispatchResult {
+	fn process_trade_fee(
+		trader: &T::AccountId,
+		asset: T::AssetId,
+		amount: Balance,
+	) -> Result<Vec<Fee<T::AccountId>>, DispatchError> {
 		let account = Self::protocol_account();
 		let original_asset_reserve = T::Currency::free_balance(asset, &account);
 
 		// Let's give little bit less to process. Subtracting one due to potential rounding errors
 		let allowed_amount = amount.saturating_sub(Balance::one());
-		let used = T::OmnipoolHooks::on_trade_fee(account.clone(), trader.clone(), asset, allowed_amount)?;
+		let fees = T::OmnipoolHooks::on_trade_fee(account.clone(), trader.clone(), asset, allowed_amount)?;
+		let additional_fees_total: Balance = fees
+			.clone()
+			.into_iter()
+			.filter_map(|opt| opt.map(|(balance, _)| balance))
+			.sum();
+
 		let asset_reserve = T::Currency::free_balance(asset, &account);
 		let diff = original_asset_reserve.saturating_sub(asset_reserve);
 		ensure!(diff <= allowed_amount, Error::<T>::FeeOverdraft);
-		ensure!(diff == used, Error::<T>::FeeOverdraft);
-		Ok(())
+		ensure!(diff == additional_fees_total, Error::<T>::FeeOverdraft);
+
+		let normal_trade_fee = amount.saturating_sub(additional_fees_total);
+
+		let mut all_trade_fees = vec![Fee {
+			asset: asset.into(),
+			amount: normal_trade_fee,
+			recipient: Self::protocol_account(),
+		}];
+
+		let additional_fee_entries: Vec<Fee<T::AccountId>> = fees
+			.iter()
+			.filter_map(|opt| {
```

Comment:
no need for filter map, just map

Reply:
since tkaen fee can contain optionals, and I wanna filter them out, filter map seems like the better option